### DISCUSSION TITLE
Isn't the whitelist downpowering pandasai a lot?

### DISCUSSION THREAD

**Discussion by mspronesti** [sinaptik-ai/pandas-ai#263] [Ideas]
The purpose of this thread is discussing what are the advantages and disadvantages of a fixed whitelist.

## Motivation
Since this feature was integrated into pandasai, a couple of threads cropped up on the discussions/issues sections asking for additional libs in the whitelist, e.g. [json+io+base64](https://github.com/gventuri/pandas-ai/discussions/166#discussioncomment-6123877), [scipy](https://github.com/gventuri/pandas-ai/discussions/235#discussion-5269201), [datetime](https://github.com/gventuri/pandas-ai/issues/260#issue-1749190516).  Needless to say, one can think of several scenarios in which other libs would be needed so just enlarging the whitelist doesn't sound like a solution to me. 

## Discussion
- What is the reason why we chose to have a whitelist of libs in first place ? 
- Can we design a way to allow any lib ?

**Comment by gventuri** [comment]
Hi @mspronesti, the reasons why we added the whitelist option is primerely 2:
* security: we don't want to risk prompt injection attacks or hallucinations to run malicious code of any kind
* scope: PandasAI is meant to be a tool to deal with data in a conversational way, so ideally we don't expect it to generate any kind of code and execute it. Given the fact that the libraries used for data analysis/data science are pretty much the same, in my opinion it's more a game of whitelisting them and then this issues shouldn't occur anymore

One thing that we could do is letting the user pass its own custom whitelisted libraries in a way that if they experience any kind of error, they can then easily have them imported.

What do you think?

**Reply by mspronesti** [reply]
that's less of a direction I'd take: we assume the user wants to perform a task in natural language, so we shouldn't assume they know which libs they require, I guess. Even for devs I believe we shouldn't: consider, for instance, the first discussion I quoted above: I ask specifically to encode the output in base64 and to provide a json, but I might not remember I also need `io` and that `base64` is the name of the lib and so on. 
In my humble opinion, if one has to provide his own appendix to the whitelist, it's easier to just google the solution/go directly to the playground of a LLM. Also, this might be frustrating: we call the API, but then the code execution fails because we forgot a dep and if this happens a couple of times we might need 3-4 calls to the API to iteratively correct the missing libs, even if the generated code is correct. What do you think ? :)

**Reply by gventuri** [reply]
What I mean is something different:
- we whitelist by default the most well known libraries for data analysis.
- we still let the user pass additional libraries if for any reason the library they need to use has not been whitelisted

The error will let them know which library is missing and the workaround to have the code running. Plus, given we have introduced the cache, no additional call will be done to the LLM.

Let me know your thoughts :)

**Reply by mspronesti** [reply]
I agree on the cache part, I didn't know about this feature. I guess it's brand new as well :) 
As regards the rest of the message, I think I had got it in first place: even if the whitelist covers the most common deps, there might be cases in which the generated code requires 2-3 extra libs and one has to manually pass them and the process becomes iterative because the execution will fail on the first disallowed import.  
I understand this is no difficult task on the user part - assuming we add a whitelist param - but I find it not much coherent with the idea that pandasai was supposed to be a "plug and play" lib also usable by non-devs[^1], doesn't it ? 
Also, if we add a webUI, this solution seems less practicable, in my opinion. 
If it is the case, I guess that we might only enrich that list hoping it covers all the scenarios :) 

[^1]:  assumption based on some previous discussions/issues on this repository.

**Reply by mspronesti** [reply]
@gventuri This [issue](https://github.com/gventuri/pandas-ai/issues/266) seems (to me) to reinforce the point that controlling the dependencies is something not really doable.

**Reply by gventuri** [reply]
@mspronesti merged an update that should improve the way we handle external libraries. Let's see if it works better now, and how likely are other errors like this to incur.

As of that issue in particular, I guess we should rather figure out a way to pass these extra libraries as additional whitelisted libraries.

Let's monitor and see what happens in the coming weeks. If we need to rollback tho, we need to find another way to enforce the security, as we don't want to expose the user using PandasAI to prompt injections or similar threats.

**Reply by gventuri** [reply]
@mspronesti it seems after the merge there aren't so many issues with whitelisted packages anymore. What do you think?
